### ISSUE TITLE
image_id incorrectly cast to int in pred_to_json, breaking image ID mapping for files like 0001.jpg

### ISSUE THREAD

**Issue by olena-hul-dataspan** [ultralytics/ultralytics#21557]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Val

### Bug

In the method `ultralytics.models.yolo.detect.val.DetectionValidator.pred_to_json`, the following line attempts to cast the image stem to an int if it's numeric:

```
image_id = int(stem) if stem.isnumeric() else stem
```

This causes unexpected behavior. For example, an image file named `0001.jpg` gets its image_id converted to 1, which breaks the mapping when comparing with the original image identifiers, especially in COCO evaluation or datasets where leading zeros are significant.

There are a couple of such occurrences in the repo https://github.com/search?q=repo%3Aultralytics%2Fultralytics%20image_id%20%3D%20int(stem)%20if%20stem.isnumeric()%20else%20stem&type=code

### Environment

Ultralytics 8.3.169 🚀 Python-3.12.1 torch-2.7.0 CPU (Apple M1 Pro)
Setup complete ✅ (10 CPUs, 16.0 GB RAM, 365.1/460.4 GB disk)

OS                  macOS-14.4.1-arm64-arm-64bit
Environment         Darwin
Python              3.12.1
Install             git
RAM                 16.00 GB
Disk                365.1/460.4 GB
CPU                 Apple M1 Pro
CPU count           10
GPU                 None
GPU count           None
CUDA                None

numpy               ✅ 1.26.4>=1.23.0
matplotlib          ✅ 3.10.3>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.12.0>=1.4.1
torch               ✅ 2.7.0>=1.8.0
torch               ✅ 2.7.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.22.0>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

```
import json
from pathlib import Path

import cv2
import numpy as np
import yaml
from ultralytics import YOLO

data_dir = Path("data")
data_dir.mkdir(exist_ok=True)

results_dir = Path("results")

# Generate 10 dummy images and labels
for i in range(10):
    img_name = f"{i:04d}"
    img_path = data_dir / f"{img_name}.jpg"
    label_path = data_dir / f"{img_name}.txt"

    # Create dummy image with white rectangle
    image = np.zeros((640, 640, 3), dtype=np.uint8)
    # Move the rectangle slightly in each image
    x1, y1 = 100 + i * 5, 150 + i * 3
    x2, y2 = x1 + 200, y1 + 200
    cv2.rectangle(image, (x1, y1), (x2, y2), (255, 255, 255), -1)
    cv2.imwrite(str(img_path), image)

    cx = ((x1 + x2) / 2) / 640
    cy = ((y1 + y2) / 2) / 640
    w = (x2 - x1) / 640
    h = (y2 - y1) / 640

    label_line = f"0 {cx:.6f} {cy:.6f} {w:.6f} {h:.6f}\n"
    label_path.write_text(label_line)

yaml_data = {
    "train": str(data_dir.resolve()),
    "val": str(data_dir.resolve()),
    "names": ["object"]
}
yaml_path = Path("dataset.yaml")
with open(yaml_path, "w") as f:
    yaml.dump(yaml_data, f, sort_keys=False)

model = YOLO("yolov8n.pt")
model.val(data=str(yaml_path), save_json=True, project=str(results_dir))

image_files_keys = [
    image.stem for image in data_dir.glob("*.jpg")
]
predictions_json = Path(results_dir / "val" / "predictions.json")
with open(predictions_json, "r") as f:
    predictions = json.load(f)

image_ids_from_predictions = {item["image_id"] for item in predictions}

print(f"Image files keys: {image_files_keys}")
print(f"Image IDs from predictions: {image_ids_from_predictions}")
```

### Additional

<img width="1230" height="290" alt="Image" src="https://github.com/user-attachments/assets/72f8a4e9-04a1-432e-81d2-8dd430b075c7" />

<img width="448" height="506" alt="Image" src="https://github.com/user-attachments/assets/74afca36-c1a1-4ac5-b16f-dff3e1fbb231" />

[predictions.json](https://github.com/user-attachments/files/21528193/predictions.json)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @olena-hul-dataspan, thank you for reporting this issue and for your detailed description and reproducible example! 🚀 This is an automated response—an Ultralytics engineer will review your report and assist you further soon.

We recommend checking our [Docs](https://docs.ultralytics.com/) for additional guidance. For bug reports like this one, providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) as you did is extremely helpful for our debugging process.

If you have further details, configuration files, or logs to add, please include them here.

Join the Ultralytics community where it suits you best:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out [Discourse](https://community.ultralytics.com/)
- To share knowledge, join our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are running the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your contribution and patience while we investigate your report! 🛠️

**Comment by Y-T-G** [comment]
In the MS-COCO dataset, the `image_id` is based on the filename cast to `int`. So that's the logic being used here. I think most labeling tools also produce `image_id` as `int` and not a string. To store leading zeros, it will have to be stored as a string.

**Comment by olena-hul-dataspan** [comment]
@Y-T-G Thanks for the clarification! It’s still a bit confusing in practice — we’re using both numeric-like filenames (e.g., 0001.jpg) and text-based ones (e.g., test.jpg). Currently, the behavior results in mixed types for image_id: an integer (1) for 0001.jpg, and a string (test) for test.jpg.

This inconsistency complicates downstream processing where uniformity in image_id types is expected — especially when matching predictions to original image files or comparing with datasets where leading zeros carry significance.

Is this type coercion (int(stem) if stem.isnumeric() else stem) the intended long-term behavior in the library? Or would it make sense to consistently treat all image_ids as strings for clarity and compatibility?

**Comment by Y-T-G** [comment]
I think it's difficult to assign correct `image_ids` without knowing how the original labels have assigned `image_ids`. In COCO convention, there's no strict connection between the filename and `image_id`. There's a separate metadata that usually tells you the mapping between filename and `image_id`. The convention used would largely depend on how the labelling tool performs the export and assigns the `image_id`.

For example, I checked Datumaro's COCO importer and the expectation in their code is that `image_id` is `int`. I think that's the expectation in general. Since Ultralytics doesn't have the mapping between filename and `image_id`, the best that can be done is to assign filename as `image_id` in case it's not numeric. In order to maintain that it's always integer, Ultralytics will have to randomly assign an integer which will confuse the user. Ultralytics can also make them all string, but then the users will have to fix the `image_ids` themselves based on the filename to `image_id` mapping in the annotation.

**Comment by Y-T-G** [comment]
Maybe we can add a separate `filename` field to the JSON which the users can use.

**Comment by olena-hul-dataspan** [comment]
@Y-T-G Hi! Adding separate `filename` field would be very useful, then we can rely on it instead of `image_id` to map predictions to original file

**Comment by Y-T-G** [comment]
Can you open a PR adding `filename` field?

**Comment by elad-shaked-dataspan** [comment]
@Y-T-G would the suggested `filename` field contain the `Path().name` or the full path `str(Path())` ?

**Comment by Y-T-G** [comment]
Just the filename

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #21837: Add `file_name` field to `predictions.json` for better image-annotation alignment. This update adds a new "file_name" key alongside "image_id" in JSON outputs from RT-DETR, YOLO Detect, and YOLO OBB validators. It improves traceability back to source images without changing model behavior or performance. Path handling was also standardized and docs/examples were clarified. See the Ultralytics Docs for details: https://docs.ultralytics.com

Big thanks to @olena-hul-dataspan for the PR and to @Laughing-q for contributions.

To test if this resolves the issue:
- Install the latest from main:
  `pip install -U "git+https://github.com/ultralytics/ultralytics.git@main"`
- Or wait for the next official PyPI release, then:
  `pip install -U ultralytics`
- Validate with JSON saving enabled and check for "file_name" in the output:
  CLI:
  `yolo val model=yolo11n.pt data=coco128.yaml save_json=True`
  Python:
  ```python
  from ultralytics import YOLO
  model = YOLO("yolo11n.pt")
  metrics = model.val(data="coco128.yaml", save_json=True)
  # Check runs/val/exp*/predictions.json for the "file_name" field
  ```

Feedback on whether this change resolves the issue would be appreciated. Thanks for reporting this, and please share any follow-ups if the problem persists. 🙏
